app.py

```diff
@@ -36,52 +48,158 @@
     r.raise_for_status()
     return r.json()
 
 
-@st.cache_resource(ttl=FIFTEEN_MINS, max_entries=1000)
-def api_data(package_name, channel="conda-forge"):
-    if channel.startswith("pkgs/"):
-        channel = channel.split("/", 1)[1]
-    r = requests.get(f"https://api.anaconda.org/package/{channel}/{package_name}/files")
+def _download_compressed_repodata(channel_name: str, arch_subdir: str) -> dict | None:
+    """
+    Try to download the compressed repodata.json file.
+    If the file does not exist, return None.
+    Other HTTP errors are raised.
+    :returns the decompressed repodata.json, or None if the compressed file does not exist
+    """
+    channel_config = get_channel_config(channel_name)
+    zstd_url = channel_config.get_zstd_repodata_url(arch_subdir)
+
+    with _make_http_session(channel_name).get(zstd_url, stream=True) as r:
+        if r.status_code == 404:
+            return None
+        r.raise_for_status()
+
+        dctx = zstd.ZstdDecompressor()
+        with dctx.stream_reader(r.raw) as reader:
+            return json.load(reader)
+
+@st.cache_resource(ttl="15m", max_entries=50)
+def get_repodata(channel_name: str, arch_subdir: str) -> dict:
+    """
+    Fetch the repodata.json for a channel and subdir.
+    This function first tries to download the .zst-compressed version of the repodata, and if that does not exist,
+    it falls back to the uncompressed.
+    """
+    decompressed_repodata = _download_compressed_repodata(channel_name, arch_subdir)
+
+    if decompressed_repodata is not None:
+        return decompressed_repodata
+
+    channel_config = get_channel_config(channel_name)
+
+    # fall back to the uncompressed version
+    repodata_url = channel_config.get_repodata_url(arch_subdir)
+    r = _make_http_session(channel_name).get(repodata_url)
     r.raise_for_status()
+
     return r.json()
 
 
-@st.cache_resource(ttl=ONE_DAY, max_entries=10)
-def repodata_patches(channel="conda-forge"):
-    package_name = f"{channel}-repodata-patches"
-    data = api_data(package_name, channel)
+def get_all_packages_sections_from_repodata(channel_name: str, arch_subdir: str, with_broken: bool) -> dict:
+    """
+    Contains the "packages" and "packages.conda" sections of the repodata.
+
+    If with_broken is True, name, version, build, and build_number of broken packages are included by performing string operations
+    on the artifact names in the "removed" section of the repodata, if exists.
+    """
+    sections = {}
+    repodata = get_repodata(channel_name, arch_subdir)
+    for key in ("packages", "packages.conda"):
+        if key in repodata:
+            sections.update(repodata[key])
+
+    if with_broken:
+        for removed_artifact in get_repodata(channel, arch_subdir).get("removed", []):
+            removed_artifact: str
+            artifact_basename, artifact_extension = removed_artifact.rsplit(".", 1)
```

Comment:
This will fail with `tar.bz2` files. You need to use something like what I posted in my previous comment, unfortunately 😬

Reply:
of course, fixed in a1f1ec1ac6fb9b84554bf4c48a1cbb475d9a8bdd